Docs/company certificate standard/update

nicoprow opened the pull request:
Pull request used for commenting and easy overview of current changes to the main. Pull request targets the fork's main branch since we do not want the current progress to be visible in the CatenaX-ev space yet.

nicoprow committed:
Update CX-0135-CompanyCertificateManagement.md

first draft for revising response codes

nicoprow committed:
Update CX-0135-CompanyCertificateManagement.md

WIP

FEILSTE committed:
Update CX-0135-CompanyCertificateManagement.md

Added fixes.

FEILSTE committed:
Added state machine image.

FEILSTE committed:
Update CX-0135-CompanyCertificateManagement.md

Fixed image link.

FEILSTE committed:
Update CX-0135-CompanyCertificateManagement.md

Fixed typo

nicoprow committed:
Update CX-0135-CompanyCertificateManagement.md

WIP

nicoprow committed:
Update CX-0135-CompanyCertificateManagement.md

WIP

nicoprow committed:
Update CX-0135-CompanyCertificateManagement.md

WIP

nicoprow committed:
Add placeholder roles for certificate provider and certificate consumer.

Adapt EDC offers.

nicoprow commented:
@FEILSTE @maximilianong 

I have added some easy to replace placeholder names for our provider and consumer roles. I also divided the offer into two offers. Its still a bit unclear to me what is the best way to go about the role definitions. I think instead of referencing the dataspace participant roles everywhere (company who provides certificates, company who wants certificates and the business solution provider), I would define two roles in terms of data exchange: certificate provider and certificate consumer. Then I would assign which dataspace participant role belongs to which data exchange role (for example business solution provider must fulfill roles of both certificate provider and certificate consumer). Then we can define what each data exchange role MUST or SHOULD do.

FEILSTE committed:
Update CX-0135-CompanyCertificateManagement.md

Updated section 2.1.5

FEILSTE committed:
Add files via upload

Updated state machine image

FEILSTE commented:
While I agree that it adds redundancy, after reading through your changes I kind of even like the placeholders. The benefit of explicitly stating the provider and consumer every time in regards of the certificate and not the technical data exchange is - in my opinion - that a developer reading this standard would always know exactly what is meant. The amount of times I did not read a standard from the top, but just did a quick search or something that directly pointed me to a subsection of the standard, by far outnumber the times I actually read it from start to finish. Someone doing a quick search, looking something up and such, I think would benefit from the explicit statement.

I also am fine with the definition of the two assets instead of three, after sleeping on it :D.

What I did now is updating the state machine image to reflect the naming changes, and I detailed out section 2.1.5 a little bit.

nicoprow committed:
replace DPPC and DPCC with Certificate Distributor and Certificate Receiver.

A few additional changes

Commit:
Adopted renaming for state machine image & fixed typo in API assets.

nicoprow committed:
add senderFeedbackUrl to feddback-endpoint request.
Explained EDC HTTP response limitations.

FEILSTE committed:
Update CX-0135-CompanyCertificateManagement.md

- Added explanation for use of Certificate Receiver/Distributor terms.
- Reworked the information regarding the feedbackUrl (also made them part of the normative standard).
- Reworked the Note for the EDC issue with 400 response codes.

FEILSTE committed:
Update CX-0135-CompanyCertificateManagement.md

Fixed formatting for one **MUST**

FEILSTE committed:
Update CX-0135-CompanyCertificateManagement.md

Fixed more normative formatting.

hschwabe commented:
> While I agree that it adds redundancy, after reading through your changes I kind of even like the placeholders. The benefit of explicitly stating the provider and consumer every time in regards of the certificate and not the technical data exchange is - in my opinion - that a developer reading this standard would always know exactly what is meant. The amount of times I did not read a standard from the top, but just did a quick search or something that directly pointed me to a subsection of the standard, by far outnumber the times I actually read it from start to finish. Someone doing a quick search, looking something up and such, I think would benefit from the explicit statement.

I agree with this: 
- People almost never read the whole document
- Be as explicit as possible when talking to people who only want to implement the function

Cvetanka-ZF commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
```diff
@@ -326,53 +415,85 @@ The API MUST use JSON formatted data transmitted over HTTPS.
-- Data Consumer MAY send a notification of reception when the certificate validation has started.
-- Data Provider MAY send a certificate availability notification when a new certificate has been published in the EDC catalog. Data Consumer SHOULD get the new certificate via the pull data exchange mechanism.
+Certificate Distributor & Certificate Receiver:
+- Certificate Distributor MUST expose company certificates in their catalog when using the pull mechanism.
```
@FEILSTE - Concerning "pull mechanism" are we also referring to the "standard" EDC catalog search and the registration of individual certificates within the catalog?
If yes, then I believe we need to provide the EDC asset structure as well. This will ensure that the Certificate Distributor understands how to correctly register the asset in EDC.
Here is an example:
```json
{
    "@id": "CompanyCertficate",
    "@type": "Asset",
    "properties": {
        "dct:certificateType": {
            "@id": "cx-taxo:ISO9001" 
        },
         "dct:enclosedSites": [
            {
                "@id": "cx-taxo:BPNS000000000001"
            },
            {
                "@id": "cx-taxo:BPNS000000000002"
            },
            {
                "@id": "cx-taxo:BPNA000000000001"
            }
         ]
        "dct:subject": {
            "@id": "cx-taxo:CompanyCertificate"
        },
        "dct:description": "Business Partner Company Certificate",
        "cx-common:version": "3.0",
    },
    "dataAddress": {
      "@type": "DataAddress",
      "type": "HttpData",
      "baseUrl": "https://backend-base-url/certificate-management-api-base-path",
      "proxyQueryParams": "true",
      "proxyPath": "false",
      "proxyMethod": "false",
      "proxyBody": "false"
    },
    "@context": {
        "dct": "http://purl.org/dc/terms/",
        "cx-taxo": "https://w3id.org/catenax/taxonomy#",
        "cx-common": "https://w3id.org/catenax/ontology/common#"
    }
}
```

FEILSTE replied:
@Cvetanka-ZF Good point, we'll add that - thank you :)

FEILSTE replied:
I added it in the latest change -> feel free to provide feedback regarding the example and the normative text beforehand that defines how the certificateType and enclosedSites are specified.

sschul33 replied:
> @FEILSTE - Concerning "pull mechanism" are we also referring to the "standard" EDC catalog search and the registration of individual certificates within the catalog? If yes, then I believe we need to provide the EDC asset structure as well. This will ensure that the Certificate Distributor understands how to correctly register the asset in EDC. Here is an example:
> 
> ```json
> {
>     "@id": "CompanyCertficate",
>     "@type": "Asset",
>     "properties": {
>         "dct:certificateType": {
>             "@id": "cx-taxo:ISO9001" 
>         },
>          "dct:enclosedSites": [
>             {
>                 "@id": "cx-taxo:BPNS000000000001"
>             },
>             {
>                 "@id": "cx-taxo:BPNS000000000002"
>             },
>             {
>                 "@id": "cx-taxo:BPNA000000000001"
>             }
>          ]
>         "dct:subject": {
>             "@id": "cx-taxo:CompanyCertificate"
>         },
>         "dct:description": "Business Partner Company Certificate",
>         "cx-common:version": "3.0",
>     },
>     "dataAddress": {
>       "@type": "DataAddress",
>       "type": "HttpData",
>       "baseUrl": "https://backend-base-url/certificate-management-api-base-path",
>       "proxyQueryParams": "true",
>       "proxyPath": "false",
>       "proxyMethod": "false",
>       "proxyBody": "false"
>     },
>     "@context": {
>         "dct": "http://purl.org/dc/terms/",
>         "cx-taxo": "https://w3id.org/catenax/taxonomy#",
>         "cx-common": "https://w3id.org/catenax/ontology/common#"
>     }
> }
> ```
this is what the asset properties look like in T-Systems implementation:

"semanticId": "io.catenax.business_partner_certificate",
                "dct:type": {
                    "@id": "https://w3id.org/catenax/taxonomy#Submodel"
                },
                "aas:semanticId": {
                    "@id": "urn:samm:io.catenax.business_partner_certificate:3.1.0#BusinessPartnerCertificate"
                },
                "name": "TestCertificate - IATF16949 - RegNo 55558",
                "id": "a6c9a5d4-1dbf-42b6-8b6d-2751d4d13cb7",
                "https://w3id.org/catenax/ontology/common#version": "3.0",
                "assettype": "Submodel",
                "contenttype": "application/json"

FEILSTE replied:
This implementation is - as of right now - not supported by the standard (as it requires asset discovery based on searching the digital twin registry - the current search mechanism is based on EDC asset properties); it is my understanding that this standard never intended to use digital twins for the certificates, but rather an API based data-exchange. If this is to change then the standard would have to be re-written in agreeance with all expert group participants (as it would require quite a large change, possible breaking - which is only doable in a major release). But as I was not part of this standard development process before the 25.09 release, someone else should confirm this (@nicoprow / @Cvetanka-ZF one of you or someone you know?)

PaddseL commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
```diff
@@ -82,29 +89,34 @@ See [API Message Flow](#215-message-flow-expectations) expectations for conforma
-- Consumer -> Provider : Company Certificate Request
-- Consumer -> Provider : Company Certificate Status (Accepted, Rejected or Received)
-- Provider -> Consumer : Company Certificate Push
-- Provider -> Consumer : Company Certificate Available
```
~The use case "Company Certificate Available" is lost here.~

EDIT: This is on purpose.

wfolkendt replied:
We have introduced the "Company Certificate Available" on purpose in April. I can NOT agree to remove the Use cCse/Application Programming Interface.

PaddseL commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
```diff
@@ -82,29 +89,34 @@ See [API Message Flow](#215-message-flow-expectations) expectations for conforma
+This request can be sent by the Certificate Receiver continuously, for updates on the request state on Certificate Distributor side.
 
- Consumer is requesting a specific certificate to Provider. Reponse could be 200 if the certificate was found or 404 otherwise.
+![alt text](assets/State%20Machine%20Certificate%20Distributor.svg "3-Tier example of a digital twin BOM")
```
Alternative text is not making sense here, copy-paste error?

PaddseL commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
```diff
@@ -113,26 +125,121 @@ We are adding the following 4 Application Programming Interfaces to remediate th
+  "header" : {
+    "senderBpn" : "BPNL0000000001AB",
+    "relatedMessageId" : "d9452f24-3bf3-4134-b3eb-68858f1b2362",
+    "context" : "CompanyCertificateManagement-CCMAPI-Status:1.0.0",
```
```suggestion
    "context" : "CompanyCertificateManagement-CCMAPI-Request:1.0.0",
```

PaddseL commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
```diff
@@ -113,26 +125,121 @@ We are adding the following 4 Application Programming Interfaces to remediate th
+  "header" : {
+    "senderBpn" : "BPNL0000000001AB",
+    "relatedMessageId" : "d9452f24-3bf3-4134-b3eb-68858f1b2362",
+    "context" : "CompanyCertificateManagement-CCMAPI-Status:1.0.0",
```
```suggestion
    "context" : "CompanyCertificateManagement-CCMAPI-Request:1.0.0",
```

PaddseL commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
```diff
@@ -113,26 +125,121 @@ We are adding the following 4 Application Programming Interfaces to remediate th
+
+Certificate is pushed by the Certificate Distributor to the Certificate Receiver.
+The enclosed BPNs can be a mix of sites and addresses.
+The Certificate Receiver may want to send a subsequent GET or fetch the asset in the catalog.
```
Why would the Certificate Receiver may want to do that? They just received the certificate via the PUSH notification.
The Certificate Receiver may want to send a feedback/status notification afterwards.

FEILSTE replied:
I think this was a re-wording of the previous sentence and this may have slipped in by accident. I'll change it to "may want to send a subsequent feedback message".

PaddseL commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
```diff
@@ -172,19 +279,30 @@ Certificate is pushed by the provider to the consumer. The enclosed Bpns can be
+>A future change is required in that regard, especially when considering the deprecation of the v1 DSP endpoint in favor of an upcoming EDC `.well-known` endpoint that supports multiple DSP versions.
 
-**Company Certificate Status**
+##### 2.1.1.3 Company Certificate Feedback
```
Renaming this only here without renaming it on other occasions (e.g. in the `context`) is just introducing more complexity and confusion.
If you plan to rename it, please make sure to do that everywhere.

FEILSTE replied:
We renamed it, because "Status" would then be confusing with the status of the request api in the pull scenario. We explicitly named it "Feedback' API, because that is what this API is for: providing feedback.

PaddseL replied:
Ok, i understand.
All in all a good change but please consider also changing other occasions (like the `context`) to avoid confusion.

FEILSTE replied:
Good point, I'll change that - that was a copy paste oversight :D

seb-l0 replied:
The renaming from `POST /companycertificate/status` to `POST /companycertificate/feedback` would, as I understand it, be a breaking change for the current solution, correct?

PaddseL commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
```diff
@@ -172,19 +279,30 @@ Certificate is pushed by the provider to the consumer. The enclosed Bpns can be
 
-**Status: Accepted**
-Certificate is accepted. Document UUID should match the incoming message. The enclosed Bpns can be a mix of sites and addresses
+This API is used by the dataspace participant who consumes a certificate to give a feedback to the certificate provider, thus either accepting or rejecting the provided certificate. This is regardless of whether the certificate was [pulled](#2111-company-certificate-request) or [pushed](#2112-company-certificate-push).
```
```suggestion
This API is used by the Certificate Receiver to give a feedback to the Certificate Distributor, thus either accepting or rejecting the provided certificate. This is regardless of whether the certificate was [pulled](#2111-company-certificate-request) or [pushed](#2112-company-certificate-push).
```

FEILSTE replied:
I'll also change it then to Certificate Provider and Certificate Consumer, based on todays voting results.

PaddseL commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
```diff
@@ -172,19 +279,30 @@ Certificate is pushed by the provider to the consumer. The enclosed Bpns can be
+
+##### 2.1.1.3.1 Company Certificate Feedback: Accepted
+
+Certificate is accepted. Document UUID should match the incoming message. The enclosed BPNs can be a mix of sites and addresses
```
```suggestion
Certificate is accepted. Document UUID should match the one from the incoming message. The enclosed BPNs can be a mix of sites and addresses
```

PaddseL commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
```diff
@@ -31,12 +38,12 @@ These enhancements are designed to improve functionality and user experience, ma
+1. Certificate Distributor wants to publish a certificate.
+2. Certificate Receiver wants to request a certificate from a specific Certificate Distributor
+3. Certificate Receiver wants to notify a Certificate Distributor of acceptance or rejection of a certificate published in #1 via a feedback message
+4. Certificate Distributor wants to notify a Certificate Receiver of the availability of a new available certificate (a new dataspace offer)
```
How would this be done without the Availability notification?

FEILSTE replied:
Maybe I'm missing it, but I think that also wasn't in before our rework. I think we just didn't consider this while re-writing. Either we have to add the notification mechanism, or we remove 4.. Perhaps this is something for a next iteration of the standard. I'd propose I remove 4. now. I'll also update 1. to include the case for the Certificate Receiver that wants to discover a provided certificate.

PaddseL commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
````diff
@@ -326,53 +415,85 @@ The API MUST use JSON formatted data transmitted over HTTPS.
+*Example Certificate Distributor API:*
+```json
+{
+  "@id": "CCMAPI",
````
This is misleading as there can't be two assets with the same assetId on one connector.
How does the setup look if we want to have both APIs configured on one connector?

FEILSTE replied:
Both APIs share the same asset, so you would not have two assets, but only one. The standard then says APIs must be able to return a 500 error; so if one of the APIs is not available (since only solution providers must provide both APIs) it would need throw a 500 when calling that API.

FEILSTE replied:
never mind - I thought you referred to the case of the Certificate Distributor

regarding the ```"@id": "CCMAPI"``` -> this should be a (dataspace participant individual / random) uuid anyways, as this could be misleading otherwise. 
Good point, we should also comment that.

Finding / distinguishing the EDC assets would be done by this:
```json
"dct:subject": {
      "@id": "cx-taxo:CompanyCertificateManagementCertificateDistributorApi"
},
```
and
```json
"dct:subject": {
      "@id": "cx-taxo:CompanyCertificateManagementCertificateReceiverApi"
},
```
respectively.

PaddseL commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
```diff
@@ -437,11 +558,11 @@ Additional more general usage policies MAY be included, but all the usage polici
-                        "operator": "eq",
-                        "rightOperand": "x12345"
-                    },
+            {
```
```suggestion
{
```

PaddseL commented:
Some other thoughts that got into my head while reading through the document:

- Section 3.2.7 TRUST LEVEL has a description for `medium` despite there being no medium trust level in the [aspect model](https://github.com/eclipse-tractusx/sldt-semantic-models/blob/main/io.catenax.business_partner_certificate/3.1.0/BusinessPartnerCertificate.ttl).
- Would it make sense to introduce asset bundling for certificates instead of creating an asset per certificate?
- Why can `locationBpns` be specified in the payload for `POST /companycertificate/feedback`? What is the intended use for it?
- If i understood it correctly the Request-Notification is not supposed to be answered by the Certificate Distributor (with a separate notification/message) but just polled by the Certificate Receiver to get the current status of it. If this is how it should be used i think this should be stated more clearly.
- An example OpenAPI YAML with the desired endpoints and data structures for this use case would make it much easier to work on an implementation.
- It's also not quite prominent/clear that the result of the Request-Notification isn't leading to a successful certificate exchange by itself. After a completed request, a subsequent PULL or PUSH has to be done afterwards.

Thanks for your time and effort in shaping this standard, i do really appreciate seeing this moving forward! :)

FEILSTE commented:
@PaddseL I added my responses to some of your comments inline:

> * Would it make sense to introduce asset bundling for certificates instead of creating an asset per certificate?

Asset bundling only works if we go with the digital twin registry / shell setup, because otherwise it would not be possible to filter for the required certificate asset. Currently each certificate has its own asset, so that the Certificate Consumer can filter for it.

> * Why can `locationBpns` be specified in the payload for `POST /companycertificate/feedback`? What is the intended use for it?

Someone with business background should explain that - we kept the changes as close to the original as possible.

> * If i understood it correctly the Request-Notification is not supposed to be answered by the Certificate Distributor (with a separate notification/message) but just polled by the Certificate Receiver to get the current status of it. If this is how it should be used i think this should be stated more clearly.

I had hoped that the State Diagram would have made that clear :D. Feel free to propose a more detailed description.

> * An example OpenAPI YAML with the desired endpoints and data structures for this use case would make it much easier to work on an implementation.

Agreed; if someone finds the time for this, that would be best.

> Thanks for your time and effort in shaping this standard, i do really appreciate seeing this moving forward! :)

You are welcome :)

FEILSTE committed:
Update CX-0135-CompanyCertificateManagement.md

Added comments from todays discussion.

FEILSTE commented:
I hope I didn't miss any of the comments so far.

FEILSTE committed:
Update CX-0135-CompanyCertificateManagement.md

Added restriction to one unique asset per API (version/subject)

nicoprow committed:
add first draft of openapi spec for certificate notification API

nicoprow committed:
remove inline ToDos and comments

Such comments should go onto the pull request on github instead

nicoprow commented on docs/standards/CX-0135-CompanyCertificateManagement/assets/openapi-spec.yaml:
```diff
@@ -0,0 +1,588 @@
+      properties:
+        senderBpn:
+          type: string
+          description: The BPNL of the legal entity which send this request
```
At the moment this is trivial but with the introduction of managed companies it is unclear whether the description is still acurate

nicoprow commented on docs/standards/CX-0135-CompanyCertificateManagement/assets/openapi-spec.yaml:
```diff
@@ -0,0 +1,588 @@
+        messageId:
+          type: string
+          format: uuid
+          description: Identifier for this specific request being sent
```
This seems to be just for the incoming message. Why not also an identifier for the actual ongoing use case process?

nicoprow commented on docs/standards/CX-0135-CompanyCertificateManagement/assets/openapi-spec.yaml:
```diff
@@ -0,0 +1,588 @@
+          enum:
+            - "CompanyCertificateManagement-CCMAPI-Push:1.0.0"
+            - "CompanyCertificateManagement-CCMAPI-Request:1.0.0"
+            - "CompanyCertificateManagement-CCMAPI-Status:1.0.0"
```
Why is that necessary? It should be clear based on the request itself. This only introduces room for bugs and misunderstandings

nicoprow commented on docs/standards/CX-0135-CompanyCertificateManagement/assets/openapi-spec.yaml:
```diff
@@ -0,0 +1,588 @@
+            - "CompanyCertificateManagement-CCMAPI-Status:1.0.0"
+        receiverBpn:
+          type: string
+          description: The intended receiver legal entity BPNL of this request
```
Similar to the sender, in a world with managed companies what is the expectation for this field?

nicoprow commented on docs/standards/CX-0135-CompanyCertificateManagement/assets/openapi-spec.yaml:
```diff
@@ -0,0 +1,588 @@
+          description: The full time and date with timezone on when this request has been sent by the sender
+          format: date-time
+          examples:
+            - "2025-05-04T00:00:00-07:00Z"
```
timezone was not specified before but makes sense

nicoprow commented on docs/standards/CX-0135-CompanyCertificateManagement/assets/openapi-spec.yaml:
```diff
@@ -0,0 +1,588 @@
+          examples:
+            - "2025-05-04T00:00:00-07:00Z"
+        version:
+          type: string
```
Version of what here exactly?

nicoprow commented on docs/standards/CX-0135-CompanyCertificateManagement/assets/OpenAPI_Spec.yaml:
```diff
@@ -0,0 +1,588 @@
+              description: Details on which areas / application types a certificate is valid for a company and/or site.
+            enclosedSites:
+              type: array
+              description: "Additional sites covered by the certificate, which can be either the Business Partner Number Site (BPNS) or Business Partner Number Address (BPNA)"
```
Why is it called site then? And in the other requests such attribute is called location. Very misleading

nicoprow committed:
add submodel reference and add openapi spec to main document.

FEILSTE commented:
Thanks @nicoprow , for adding the openAPI <3
*Note*: The aspect model is now also directly linked in the openAPI document (the auto-generated schema) -> if there are version changes to the aspect model, the ```$ref``` to the sldt-semanitc-models repository must be updated in the openAPI document as well, not "only" the version number in the standard document: 
- here: <img width="1378" height="203" alt="image" src="https://github.com/user-attachments/assets/2f7d1585-da5c-4708-ab42-eb9555772b7a" />

- and here: <img width="1125" height="219" alt="image" src="https://github.com/user-attachments/assets/95d11239-d07e-44fe-a4fd-d1937e9baf4c" />

Commit:
Update based on todays expert group discussion

- Added all changes from todays expert group discussion
- refactored naming
- updated all legacy images and changed them from png to svg

hschwabe commented:
@Cvetanka-ZF 

Not sure if we have a PR open for the data model as well. So, I dump the issue here, sorry!


In the semantic data model schemas, I couldnt find the parameter "language" (default = en?). Cross comparison with Certificate of Analysis reveals that a parameter "language" is indeed useful.

We should discuss this in the next call: 

 https://github.com/eclipse-tractusx/sldt-semantic-models/blob/main/io.catenax.material.certificate_of_analysis/1.0.0/gen/CertificateOfAnalysis.json

Cvetanka-ZF commented:
> @Cvetanka-ZF
> 
> Not sure if we have a PR open for the data model as well. So, I dump the issue here, sorry!
> 
> In the semantic data model schemas, I couldnt find the parameter "language" (default = en?). Cross comparison with Certificate of Analysis reveals that a parameter "language" is indeed useful.
> 
> We should discuss this in the next call:
> 
> https://github.com/eclipse-tractusx/sldt-semantic-models/blob/main/io.catenax.material.certificate_of_analysis/1.0.0/gen/CertificateOfAnalysis.json

@hschwabe 
Yes, you're right — the CCM data model does not currently include a dedicated data attribute for the language property. 
By default, English is used, and this is only reflected in the attributes descriptions within the .ttl file.
If there is a need for this to be explicitly modeled, we can discuss it in the Expert group and  consider extending the data model accordingly.

PaddseL commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
```diff
@@ -82,29 +88,36 @@ See [API Message Flow](#215-message-flow-expectations) expectations for conforma
+
+Use cases to provide certificates (from who initiates communication, negotiation and data transfer):
+
+- Certificate Consumer -> Certificate Provider : Company Certificate Request
```
nit: technically the Company Certificate Request itself does not provide certificates, just the information if one is available and ready to be pulled. A Request can also be declined, not resulting in a certificate exchange.

```suggestion
- Certificate Consumer -> Certificate Provider : Company Certificate Pull
```

However, the pulling of a certificate can be done by itself without the need of a Request-notification.

nicoprow replied:
Fair enough I took over this change

PaddseL commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
```diff
@@ -82,29 +88,36 @@ See [API Message Flow](#215-message-flow-expectations) expectations for conforma
+This request can be sent by the Certificate Consumer continuously, for updates on the request state on Certificate Provider side.
 
- Consumer is requesting a specific certificate to Provider. Reponse could be 200 if the certificate was found or 404 otherwise.
+![alt text](images/state-machine-certificate-distributor.svg "Certificate Request API State Machine")
```
The descriptions of the HTTP response codes in the table from `2.1.1.1.1 HTTP Response Codes` and the Status Codes in the state machine do not align ("Request IN PROGRESS" on the bottom left in the state machine is a 200 while the table states that the 200 is only returned for a completed request).

nicoprow replied:
True, I changed it to 202 in the figure

PaddseL commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
```diff
@@ -82,29 +88,36 @@ See [API Message Flow](#215-message-flow-expectations) expectations for conforma
+This request can be sent by the Certificate Consumer continuously, for updates on the request state on Certificate Provider side.
 
- Consumer is requesting a specific certificate to Provider. Reponse could be 200 if the certificate was found or 404 otherwise.
+![alt text](images/state-machine-certificate-distributor.svg "Certificate Request API State Machine")
```
The state transition from "Request is new & accepted: Status Code 202" points at the wrong state in my opinion - as this is returning a status code, it should point to the Start State on the (currently) bottom left of the graph because from there on a new request to identify the current status can be initiated.

nicoprow replied:
From a very strict point of view changing the transition is also not entirely correct, as in each state the process could undergo state transitions without the certificate consumer making a request. I think the idea of the state machine is clear but maybe needs to be a bit strictor formalized @FEILSTE ?

PaddseL commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
```diff
@@ -82,29 +88,36 @@ See [API Message Flow](#215-message-flow-expectations) expectations for conforma
+This request can be sent by the Certificate Consumer continuously, for updates on the request state on Certificate Provider side.
 
- Consumer is requesting a specific certificate to Provider. Reponse could be 200 if the certificate was found or 404 otherwise.
+![alt text](images/state-machine-certificate-distributor.svg "Certificate Request API State Machine")
```
Identification of a Request-notification based on basically all of its attributes (senderBpn, receiverBpn, certifiedBpn, certificateType, locationBpns) is not ideal performance-wise, a single notificationId would be better.

nicoprow replied:
I agree but this creates a more complex interaction where an additional endpoint is needed. For the sake of not introducing many changes we kept it to these combined values instead to match certificate requests.

PaddseL commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
```diff
@@ -113,26 +126,126 @@ We are adding the following 4 Application Programming Interfaces to remediate th
+{
+  "header" : {
+    "senderBpn" : "BPNL0000000001AB",
+    "relatedMessageId" : "d9452f24-3bf3-4134-b3eb-68858f1b2362",
```
The example for relatedMessageId is a little bit confusing because there is no message with such an ID anywhere in this or other examples of the standard. As this is an optional field, better leave this out (here and on other similar occasions).
```suggestion
```

nicoprow replied:
I'm more than happy to remove this field but this constitutes a breaking change. Do we know that nobody is relying on it?

sschul33 replied:
It should be removed. This is the initial message within the flow which means there cannot be a relatedMessageId.

PaddseL commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
```diff
@@ -113,26 +126,126 @@ We are adding the following 4 Application Programming Interfaces to remediate th
+    "senderBpn" : "BPNL0000000001AB",
+    "relatedMessageId" : "d9452f24-3bf3-4134-b3eb-68858f1b2362",
+    "context" : "CompanyCertificateManagement-CCMAPI-Request:1.0.0",
+    "messageId" : "3b4edc05-e214-47a1-b0c2-1d831cdd9ba9",
```
The examples are all reusing the same `messageId` which might indicate that the ID should be reused, but from my understanding there should be a new `messageId` for every individual HTTP request. I would suggest to also use different IDs for the examples.

nicoprow replied:
I set a different UUID for each messageID and relatedMessageID value

PaddseL commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
```diff
@@ -113,26 +126,126 @@ We are adding the following 4 Application Programming Interfaces to remediate th
+
+Finished Processing and Certificate Available in EDC.
+The content body also provides the documentId of the certificate.
+This allows for skipping the EDC catalog search.
```
`This allows for skipping the EDC catalog search.`

This is from my understanding a wrong statement because we need the Asset ID, the Offer ID AND the permissions to be able to initiate a EDR token negotiation. Even if the Asset ID would be enough there is no guarantee that this is the same as the document ID - there is no written requirement to use the document ID as Asset ID.

nicoprow replied:
That is true. The documentID is the assetID and merely helps with finding the correct offer in the catalog. I adapted this information accordingly

PaddseL commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
```diff
@@ -172,21 +285,35 @@ Certificate is pushed by the provider to the consumer. The enclosed Bpns can be
+The ´senderFeedbackUrl´ specifies, where the Certificate Provider expects the feedback of the Certificate Consumer.
+The expected value **MUST** be a concrete path to the version 1 dataspace protocol endpoint (as specified in the example above),
+where a data offer for an asset of type cx-taxo:CCMAPI **MUST** be available for the Certificate Consumer.
+
```
nit: A hint that the Push-Notification is the only notification where the attribute `documentID` is spelled with a capital D at the end (due to the spelling in the aspect model) might be good here. This is prone to errors when implementing.

nicoprow replied:
I added such a hint

PaddseL commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
````diff
@@ -172,21 +285,35 @@ Certificate is pushed by the provider to the consumer. The enclosed Bpns can be
 }
 ```
+The ´senderFeedbackUrl´ specifies, where the Certificate Provider expects the feedback of the Certificate Consumer.
+The expected value **MUST** be a concrete path to the version 1 dataspace protocol endpoint (as specified in the example above),
````
nit: This here is a normative MUST requirement which is referencing a non-normative example.

nicoprow replied:
I removed the remark in the paranthesis

PaddseL commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
```diff
@@ -172,21 +285,35 @@ Certificate is pushed by the provider to the consumer. The enclosed Bpns can be
+This API is used by the Certificate Consumer to give a feedback to the Certificate Provider, thus either accepting or rejecting the provided certificate.
+This is regardless of whether the certificate was [pulled](#2111-company-certificate-request) or [pushed](#2112-company-certificate-push).
+
+##### 2.1.1.3.1 Company Certificate Feedback: Accepted
```
nit: Ordering is strange here, i would expect the `Received` status to be the first one because the `Received` status only makes sense to return before `Accepted` or `Rejected`.

nicoprow replied:
Agreed, reordered

PaddseL commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
```diff
@@ -172,21 +285,35 @@ Certificate is pushed by the provider to the consumer. The enclosed Bpns can be
+##### 2.1.1.3.1 Company Certificate Feedback: Accepted
+
+Certificate is accepted. 
+The documentId should match the documentId from the incoming message. 
```
`The documentId should match the documentId from the incoming message.`
1. `should` not written in all-capital-letters, is this intended to be non-normative?
2. There might not be an incoming message if the certificate was pulled without previous Request-notification.

nicoprow replied:
I think this should be a MUST and worded it now like that. I also rephrased what the expected value of the documentId should be be.

PaddseL commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
```diff
@@ -172,21 +285,35 @@ Certificate is pushed by the provider to the consumer. The enclosed Bpns can be
+
+Certificate is accepted. 
+The documentId should match the documentId from the incoming message. 
+The enclosed BPNs can be a mix of sites and addresses.
```
```suggestion
The `locationBpns` can be a mix of sites and addresses.
```

nicoprow replied:
Done

PaddseL commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
```diff
@@ -320,37 +418,140 @@ The following http response codes MUST be defined for all resources:
 > *This section is normative*
 
-The HTTP endpoints introduced in chapter 2.1.1 API endpoints and resources MUST NOT be called from a business partner directly. Rather, it MUST be called via an connector communication. Therefore, exactly one asset MUST be offered in the connector catalog.
+The HTTP endpoints introduced in chapter [2.1.1 API endpoints and resources](#211-API-endpoints-and-resources) **MUST NOT** be called from a business partner directly. Rather, it **MUST** be called via a connector communication.
```
```suggestion
The HTTP endpoints introduced in chapter [2.1.1 API endpoints and resources](#211-API-endpoints-and-resources) **MUST NOT** be called from a business partner directly. Rather, it **MUST** be called with any CX-0018 compliant connector via dataspace protocol.
```

nicoprow replied:
Done

PaddseL commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
```diff
@@ -320,37 +418,140 @@ The following http response codes MUST be defined for all resources:
+
+| **Type**       | **Subject**                              | **Version** | **Description**                                                                                                                                                                                                |
+|----------------|------------------------------------------|-------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| cx-taxo:CCMAPI | cx-taxo:CompanyCertificateDistributorApi | 3.0         | Offers *Certificate Provider API* for the Certificate Consumer to [request](#2111-company-certificate-request) certificates and send [feedback](#2113-company-certificate-feedback) for provided certificates. |
```
Why is a different naming introduced here?
In my opinion this just adds more complexity, sticking to the already used terms would be better.
```suggestion
| cx-taxo:CCMAPI | cx-taxo:CompanyCertificateProviderApi | 3.0         | Offers *Certificate Provider API* for the Certificate Consumer to [request](#2111-company-certificate-request) certificates and send [feedback](#2113-company-certificate-feedback) for provided certificates. |
```

nicoprow replied:
Leftover from previous terms. I changed it to `Provider`

PaddseL commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
```diff
@@ -320,37 +418,140 @@ The following http response codes MUST be defined for all resources:
+| **Type**       | **Subject**                              | **Version** | **Description**                                                                                                                                                                                                |
+|----------------|------------------------------------------|-------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| cx-taxo:CCMAPI | cx-taxo:CompanyCertificateDistributorApi | 3.0         | Offers *Certificate Provider API* for the Certificate Consumer to [request](#2111-company-certificate-request) certificates and send [feedback](#2113-company-certificate-feedback) for provided certificates. |
+| cx-taxo:CCMAPI | cx-taxo:CompanyCertificateReceiverApi    | 3.0         | Offers *Certificate Consumer API* for the Certificate Provider to [push](#2112-company-certificate-push) certificates to the Certificate Consumer.                                                             |
```
```suggestion
| cx-taxo:CCMAPI | cx-taxo:CompanyCertificateConsumerApi    | 3.0         | Offers *Certificate Consumer API* for the Certificate Provider to [push](#2112-company-certificate-push) certificates to the Certificate Consumer.                                                             |
```

nicoprow replied:
Leftover from previous terms, I changed it to `Consumer`

PaddseL commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
```diff
@@ -320,37 +418,140 @@ The following http response codes MUST be defined for all resources:
 
+
+
+*Example Certificate Provider API:*
```
nit: Introducing sub-sections for each individual asset example would make it easier to recognize which text-block is belonging to which example.

nicoprow replied:
Restructured the section a bit and added subsections for clarity

PaddseL commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
```diff
@@ -320,37 +418,140 @@ The following http response codes MUST be defined for all resources:
     "properties": {
         "dct:type": {
-            "@id": "cx-taxo:CCMAPI"
+            "@id": "https://w3id.org/catenax/taxonomy#Submodel"
```
```suggestion
            "@id": "cx-taxo:Submodel"
```

nicoprow replied:
Done

PaddseL commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
```diff
@@ -320,37 +418,140 @@ The following http response codes MUST be defined for all resources:
+      "@type": "DataAddress",
+      "type": "HttpData",
+      "baseUrl": "https://backend-base-url/certificate-management-api-base-path",
+      "proxyQueryParams": "true",
```
Why is `proxyQueryParams` set to true? AFAIK we don't utilize them when pulling certificates.

```suggestion
      "proxyQueryParams": "false",
```

nicoprow replied:
As far as I know query parameters are not needed here. Changed to false

PaddseL commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
```diff
@@ -359,27 +560,43 @@ The property [[type]](http://purl.org/dc/terms/type) MUST reference the name of
+Certificate Provider & Certificate Consumer:
+- Certificate Provider **MUST** expose company certificates in their catalog when using the pull mechanism.
+- Certificate Provider **MUST** set the correct access policy on the certificate offer to allow consumption by Consumer(s) when using the pull mechanism.
+
```
What about the Certificate Provider API? Is this a **MUST** requirement?

PaddseL commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
```diff
@@ -359,27 +560,43 @@ The property [[type]](http://purl.org/dc/terms/type) MUST reference the name of
+![PULL Scenarios](images/certificate-pull.svg)
 
 The Certificate PULL Diagram describes the process of Consumer retrieving a certificate from a Provider via an EDC.
 It begins with the provider creating a Certificate Asset with corresponding contract definition in the EDC Catalog. The Consumer searches the catalog using specific filters, initiates a contract negotiation, and retrieves the Endpoint Data Reference (EDR). The Data Plane then facilitates secure data transfer, allowing the consumer to pull the certificate. Once retrieved, the Backend Certificate C ...
```
The whole sub-section `2.1.5.3 PUSH notification followed by PULL mechanism` doesn't make any sense to me, both mechanisms are for consuming certificates, therefore you can use one OR the other, but one following the other doesn't make sense to me (for normal cases, one potential use case could be an update of a certificate).

Also, there are no `Certificate Available Notification`s anymore which are mentioned there.

nicoprow replied:
Looks to me like a legacy section we forgot to adapt. I also didn't see relevance now and removed it.

PaddseL commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
```diff
@@ -390,13 +607,15 @@ After the data provider has created a Certificate Asset with corresponding contr
 
-The “Connector Asset for Certificate Notifications” included in the EDC of a data consumer MUST contain a usage policy that includes the Catena-X Data Exchange Governance document in the latest version.
-That includes the following usage purpose:
+All assets included in the EDC of a dataspace participant (APIs and Certificates) **MUST** contain a usage policy that includes the Catena-X Data Exchange Governance document in the latest version.
```
nit: Might not be a good idea to force the latest version as newer policy versions might not be compatible. Better define a fixed version, the current latest one.

nicoprow replied:
Makes sense to me. I added 1.0 version as this is the current one.

PaddseL commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
````diff
@@ -390,13 +607,15 @@ After the data provider has created a Certificate Asset with corresponding contr
+*Minimal example of a usage policy without contract reference:*
 ``` json
 {
   "@context": [
````
The `@context` example here looks strange to me, i haven't seen it like this before. I'm not sure if this can be an array.

nicoprow replied:
This is actually legit and working

PaddseL commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
````diff
@@ -434,14 +653,14 @@ Additional more general usage policies MAY be included, but all the usage polici
 ```
 
-The left operand "leftOperand": "cx-policy:ContractReference" MUST be included only if such a bilateral framework contract exists.
+The left operand "leftOperand": "cx-policy:ContractReference" **MUST** be included only if such a bilateral framework contract exists.
````
```suggestion
The constraint `{ "leftOperand": "cx-policy:ContractReference" }` **MUST** be included only if such a bilateral framework contract exists.
```

nicoprow replied:
Applied the changes

PaddseL commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
````diff
@@ -434,14 +653,14 @@ Additional more general usage policies MAY be included, but all the usage polici
+}
 ```
 
 ## 3 ASPECT MODELS
````
nit: Add links to mentions of `[CX-0003:1.1]`

nicoprow replied:
Added the links

PaddseL reviewed (changes requested):
Thanks again for your efforts, please take all comments and suggestions with a grain of salt.

One thing i would like to highlight is using the `certificateType` directly in the Asset like it is shown in the changes.
Currently, the certificate type is a free-text field, where every participant or user might have a different wording of the same type (this is explicitly stated in 3.2.2 CERTIFICATE TYPE: `Note: The spelling of the certificate type may vary slightly on the user interface or within the data model.`).
Using it as an attribute in the asset and therefore as filter is prone to errors on user-side as different implementations might be "incompatible" when trying to exchange certificates for the same type with different spellings.

On the other hand, the aspect model explicitly wants this to be a free-text field because, according to the description, `valid types are registered at BPN metadatacontroller` which forces this to be a free-text field.

One suggestion could be to force a spelling for known certificate types in a normative section but leave it open for non-known. Or force a generic spelling like `has to be spelled all lowercase without whitespaces, underscores, dashes or such`.

nicoprow committed:
incorporate review remarks

nicoprow commented:
> Thanks again for your efforts, please take all comments and suggestions with a grain of salt.
> 
> One thing i would like to highlight is using the `certificateType` directly in the Asset like it is shown in the changes. Currently, the certificate type is a free-text field, where every participant or user might have a different wording of the same type (this is explicitly stated in 3.2.2 CERTIFICATE TYPE: `Note: The spelling of the certificate type may vary slightly on the user interface or within the data model.`). Using it as an attribute in the asset and therefore as filter is prone to errors on user-side as different implementations might be "incompatible" when trying to exchange certificates for the same type with different spellings.
> 
> On the other hand, the aspect model explicitly wants this to be a free-text field because, according to the description, `valid types are registered at BPN metadatacontroller` which forces this to be a free-text field.
> 
> One suggestion could be to force a spelling for known certificate types in a normative section but leave it open for non-known. Or force a generic spelling like `has to be spelled all lowercase without whitespaces, underscores, dashes or such`.

In the certificate type section I know added rules for the spelling of certificate type values. I also added a code table resulting from these spelling rules for all currently supported certificate types.

nicoprow committed:
merge EDC Assets for Certificate Notification API

nicoprow committed:
rewording message flow expectations to match one API asset

nicoprow committed:
fix wordings for certificate type and merging of notification API

seb-l0 commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
I would like to keep the attribute: `"senderFeedbackUrl": "https://domain.tld/path/to/api",` as **optional** for all three feedback notifications (Received, Accepted, Rejected).
This would avoid adjustments and retesting.

PaddseL replied:
From my understanding there can't be a response to a feedback notification, therefore making the `senderFeedbackUrl` obsolete for these payloads.

seb-l0 commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
CTPAT | ctpat ist missing.

wfolkendt commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
```diff
@@ -31,12 +38,11 @@ These enhancements are designed to improve functionality and user experience, ma
-1. Data Provider wants to publish a certificate.
-2. (*new*) Data Consumer wants to request a certificate from a specific Data Provider
-3. (*new*) Data Consumer wants to notify a Data Provider of acceptance or rejection of a certificate published in #1 via a status message
-4. (*new*) Data Provider wants to notify a Data Consumer of the availability of a new Asset
```
We have added this entry text 6 month ago to enable the notification of a certificate consumer by the certificate provider that a new certificate is available. As a resulting action the certificate consumer can "pull" the certificate. Until now the certificate consumer has to periodically "poll" for required certificates. 
This feature is needed for certificate consumer who request data via the "pull" mechanism.
I can NOT agree to remove this option!

wfolkendt replied:
In the last regular meeting we decided to describe the business benefit of the "Certificate Available" notification. Based on this description and the following discussion we will decide related the above change.
Today some certificate consumers  use the "pull" and others use the "push" mechanism. We have agreed that both mechanism are needed and should be supported.
Certificate consumers who use the "pull" mechanism have the problem, that if a certificate is not available on provider side, then the consumer has to "poll", that means he has to request the certificate periodically and this creates effort and costs.
This problem can be solved if the certificate provider notifies the consumer if a new certificate instance is available (e.g. the successor of an existing expired certificate). The notification should contain all the data that are required for the consumer to request a certificate from the provider via the "pull" mechanism. The consumer will be interested to use this notification as a trigger to request the certificate instance via the pull mechanism.
The business benefit for the provider (as a supplier) is, that he can fullfill his delivery contract obligations to deliver company certificates on time to the consumer (as a customer).
The business benefit for the consumer is, that he does not need to configure the EDC with the BPNL specific entries referencing the bilateral contracts. Bigger companies have several thousands of providers and therefore the configuration effort is significant.

wfolkendt commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
```diff
@@ -82,59 +88,166 @@ See [API Message Flow](#215-message-flow-expectations) expectations for conforma
-- Consumer -> Provider : Company Certificate Request
-- Consumer -> Provider : Company Certificate Status (Accepted, Rejected or Received)
-- Provider -> Consumer : Company Certificate Push
-- Provider -> Consumer : Company Certificate Available
```
We have added this entry text 6 month ago to enable the notification of a certificate consumer by the certificate provider. I can NOT agree to remove this option!

wfolkendt commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
```diff
@@ -359,44 +547,54 @@ The property [[type]](http://purl.org/dc/terms/type) MUST reference the name of
 
-- Data Provider MUST expose company certificates in their catalog when using pull mechanism.
-- Data Provider MUST set the correct access policy on the certificate asset to allow consumption by Consumer(s) when using pull mechanism.
-- Data Consumer MUST set the correct access policy on the notification policy to allow the exchange of data by provider when using push mechanism.
```
Why does this statement was removed?
Assumption: When the push mechanism is used, the certificate provider needs access to the EDC of the certificate consumer to provide the certificate data.
What is the difference between "setting an access policy on the notification policy" and setting an "access policy on the offer" by the data consumer?

wfolkendt replied:
I have got the answer from the Data Sovereignity EG.  "access policies on the offer" is the correct formulation.
However this leads to a required correction in chapter 2.1.6
Current text: 
"2.1.6 Usage Policy
The “Connector Asset for Certificate Notifications” included in the EDC of a data consumer MUST contain a usage policy that includes the Catena-X Data Exchange Governance document in the latest version. That includes the following usage purpose:
•	cx.ccm.base:1: The exchanged business partner certificates are used for the purpose of verification of the existence of a certification. As a purpose-specific requirement, the duration of (i) contract, (ii) data provision and (iii) usage right(s) as a default are all specified as 1 year.
Additional more general usage policies MAY be included, but all the usage policies MUST contain the above usage purpose."

We have to agree to the new text. The following changes are required:
1. The "Connector Asset for Certificate Notification" has to removed and replaces with "EDC" respective "EDC offer". I do not know the correct term for the data structure in the EDC that contains the usage purpose and the contract reference.
2. "MUST contain a usage policy that includes the Catena-X Data Exchange Governance document in the latest version. " is a contradiction to the statement that "Version 1.0" of the usage purpose should apply made in the changes above. The contradiction needs to be removed.

wfolkendt commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
```diff
@@ -359,44 +547,54 @@ The property [[type]](http://purl.org/dc/terms/type) MUST reference the name of
-- Data Provider MAY send a certificate availability notification when a new certificate has been published in the EDC catalog. Data Consumer SHOULD get the new certificate via the pull data exchange mechanism.
+Certificate Provider & Certificate Consumer:
+- Certificate Provider **MUST** support at least one of the certificate provision mechanisms (push and/or pull mechanism)
+- Certificate Provider **MUST** expose company certificates in their catalog when using the pull mechanism.
```
What does "expose company certificates" means?

PaddseL replied:
In this context this means that in order for a receiver to consume certificates via the PULL-mechanism from the provider, the provider must ensure that for every consumable certificate there has to be a corresponding asset in their catalog.

wfolkendt commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
```diff
@@ -82,59 +88,166 @@ See [API Message Flow](#215-message-flow-expectations) expectations for conforma
+
+![alt text](images/state-machine-certificate-distributor.svg "Certificate Request API State Machine")
 
 `POST /companycertificate/request`
```
I was asked to check if the notification mechanismm used in CCM is compliant to CX_0151_Industry Core Basics. The following example is provided in this standard: 1.4 EXAMPLES
Notifcation in JSON format
{
  "header": {
    "messageId": "urn:uuid:f9a97301-a000-44dd-b9d8-78488a40c6bb",
    "context": "IndustryCore-DigitalTwinEventAPI-ConnectToParent:3.0.0",
    "sentDateTime": "2024-07-05T08:13:33.20733Z",
    "senderBpn": "BPNL000000000AAA",
    "receiverBpn": "BPNL000000000ZZZ",
    "expectedResponseBy": "2024-07-08T08:13:33.20733Z",
    "version": "3.0.0"
  },
  "content": {}
}

Does the following request notification is compliant to the above standard?
The "expectedResponseBy": ... is missing? Do we need to add something?

PaddseL replied:
Examples in general are non-normative and should not be treated as strict criteria

The `expectedResponseBy` is an optional attribute in the notification API from the CX 0151 Industry Core Basics standard, but also missing completely in the example OpenAPI specification here.

I think it would be a good idea to add it as optional attribute to the example OpenAPI specification, but not to the examples in the standard description itself.

wfolkendt commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
```diff
@@ -359,44 +547,54 @@ The property [[type]](http://purl.org/dc/terms/type) MUST reference the name of
 
 ##### 2.1.6 Usage Policy
 
-The “Connector Asset for Certificate Notifications” included in the EDC of a data consumer MUST contain a usage policy that includes the Catena-X Data Exchange Governance document in the latest version.
```
What is the benefit of the reference to version 1.0 instead the reference to "the latest version"?

PaddseL replied:
Quoting my comment from a previous review:
> nit: Might not be a good idea to force the latest version as newer policy versions might not be compatible. Better define a fixed version, the current latest one.

wfolkendt commented on docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md:
```diff
@@ -359,44 +547,54 @@ The property [[type]](http://purl.org/dc/terms/type) MUST reference the name of
+- Certificate Provider **MUST** expose company certificates in their catalog when using the pull mechanism.
+- Certificate Provider **MUST** set the correct access policy on the certificate offer to allow consumption by Consumer(s) when using the pull mechanism.
+
+- Certificate Consumer **MAY** implement the [push endpoint](#2112-company-certificate-push) for the Certificate Provider to push certificates to, but **MUST** set the correct access policy on the offer, when choosing to do so.
```
It is mandatory that certificate consumers"... **MUST* set the correct access policy **and also the correct bilateral contract reference**..." Please change the corresponding sentence above.

PaddseL replied:
Please take a look at line 635:

> The constraint `{ "leftOperand": "cx-policy:ContractReference" }` **MUST** be included only if such a bilateral framework contract exists.

Is this what you are referencing?

nicoprow commented:
outdated
